### ISSUE TITLE
Create API address derivation path type.

### ISSUE THREAD

**Issue by jonathanknowles** [cardano-foundation/cardano-wallet#2174]
# Context

We need to be able to represent address derivation paths in the API.

# Decision

Create type `ApiAddressDerivationPath` that can encode an address derivation path.

A derivation path should be serialized to JSON as a **list**.

For example, the path `1852H/1815H/0H/0/0` should have the following JSON representation:

```json
[
  {
    "derivation_index": 1852,
    "derivation_type": "hardened"
  },
  {
    "derivation_index": 1815,
    "derivation_type": "hardened"
  },
  {
    "derivation_index": 0,
    "derivation_type": "hardened"
  },
  {
    "derivation_index": 0,
    "derivation_type": "soft"
  },
  {
    "derivation_index": 0,
    "derivation_type": "soft"
  }
]
```

Each entry in the list should have the following pair of fields:

| Field name | Type | Domain |
| -- | -- | -- |
| `derivation_index` | natural number | `[ 0 .. 2^31 − 1 ]` |
| `derivation_type` | enumeration | `[ hardened \| soft ]` | 

# Acceptance Criteria

- [x] There **must** be an `ApiAddressDerivation` type.
- [x] Values of the `ApiAddressDerivation` type **must** be serializable to JSON and deserializable from JSON.
- [x] Deserialization of a previously serialized value **must** yield the original value.
- [x] The JSON representation of an `ApiAddressDerivation` value **should** be a list.

---

# Development 

- https://github.com/input-output-hk/cardano-wallet/pull/2177

# QA 

See https://github.com/input-output-hk/cardano-wallet/pull/2177, which includes:
- JSON roundtrip encoding/decoding tests for all types.
- JSON golden tests for all types.
- negative tests for the `ApiRelativeAddressIndex` type.

**Comment by paweljakubas** [comment]
I wonder if first three paths, ie., are not superfluous 1852H/1815H/0H. the first two for sure if we are dealing with ada. Last one depicting account maybe also... On the other hand, it is only presentation layer so maybe the full information does not hurt.

Another thing : we want to use relative indexes (counting from 0) also for hardened index?

**Comment by rvl** [comment]
Was there anything wrong with just using a string to represent derivation path?
It's pretty simple to parse.

**Comment by jonathanknowles** [comment]
> Was there anything wrong with just using a string to represent derivation path?
> It's pretty simple to parse.

Hi @rvl.

We could indeed have used a string, and you're right, it would not be very hard to parse.

The use of a JSON list was requested here: https://input-output-rnd.slack.com/archives/GBT05825V/p1600859749018000?thread_ts=1600843830.004300&cid=GBT05825V

**Comment by jonathanknowles** [comment]
@paweljakubas 
> Another thing : we want to use relative indexes (counting from 0) also for hardened index?

Good point! I've updated the task description to specify that we will use relative indices, ranging from `0` to `2^31 - 1`.

**Comment by rvl** [comment]
In that slack thread, a JSON list of strings was requested, not a list of objects. I fail to see how a list of objects like this will help the user.

**Comment by piotr-iohk** [comment]
lgtm
